src/components/icon/icon.ts

```diff
@@ -87,46 +85,62 @@ export default class SlIcon extends ShoelaceElement {
     const library = getIconLibrary(this.library);
     const url = this.getUrl();
 
-    // Create an instance of the DOM parser. We do it here instead of top-level to support SSR while maintaining a
-    // single parser instance for optimal performance.
-    if (!parser) {
-      parser = new DOMParser();
+    if (!url) {
+      this.svg = null;
+      return;
     }
 
-    if (url) {
-      try {
-        const file = await requestIcon(url);
-        if (url !== this.getUrl()) {
-          // If the url has changed while fetching the icon, ignore this request
-          return;
-        } else if (file.ok) {
-          const doc = parser.parseFromString(file.svg, 'text/html');
-          const svgEl = doc.body.querySelector('svg');
-
-          if (svgEl !== null) {
-            svgEl.part.add('svg');
-            library?.mutator?.(svgEl);
-            this.svg = svgEl.outerHTML;
-            this.emit('sl-load');
-          } else {
-            this.svg = '';
-            this.emit('sl-error');
-          }
-        } else {
-          this.svg = '';
-          this.emit('sl-error');
-        }
-      } catch {
+    let iconResolver = iconCache.get(url);
+    if (!iconResolver) {
+      iconResolver = SlIcon._resolveIcon(url);
+      iconCache.set(url, iconResolver);
+    }
+
+    try {
+      const svg = await iconResolver;
+      if (url !== this.getUrl()) {
+        // If the url has changed while fetching the icon, ignore this request
+        return;
+      }
+
+      if (!svg) {
+        this.svg = null;
         this.emit('sl-error');
+        return;
       }
-    } else if (this.svg.length > 0) {
-      // If we can't resolve a URL and an icon was previously set, remove it
-      this.svg = '';
+
+      this.svg = svg.cloneNode(true) as SVGElement;
+      library?.mutator?.(this.svg);
+      this.emit('sl-load');
+    } catch {
+      this.svg = null;
+      iconCache.delete(url);
+      this.emit('sl-error');
     }
   }
 
   render() {
-    return html` ${unsafeSVG(this.svg)} `;
+    return this.svg;
+  }
+
+  private static async _resolveIcon(url: string): Promise<SVGSVGElement | null> {
+    const fileData = await fetch(url, { mode: 'cors' });
+    if (!fileData.ok) return null;
+
+    const div = document.createElement('div');
+    div.innerHTML = await fileData.text();
+
+    const svg = div.firstElementChild;
+    if (svg?.tagName?.toLowerCase() !== 'svg') return null;
+
+    if (!parser) parser = new DOMParser();
+    const doc = parser.parseFromString(svg.outerHTML, 'text/html');
```

Comment:
I've done this because this is exactly what `sl-icon` was previously doing (but split across two files).

However, I'm not sure _why_ the HTML is added to the div then re-parsed again. Simply skipping that step and doing this seems to work perfectly fine:
```suggestion
    if (!parser) parser = new DOMParser();
    const doc = parser.parseFromString(await fileData.text(), 'text/html');
```

I've left this as a suggestion instead of committing it because I don't know the history around this and maybe it somehow fixes some weird edge case, but it looks to me like this can be simplified.

Reply:
No weird edge cases that I can think of, but we want to make sure that 1) we're working with an actual SVG, 2) we're adding the `svg` part to it, and 3) we're running the mutator function and passing it the same SVG. Otherwise, it should be fine.